In [2]:
text = '''Mahabharata, (Sanskrit: “Great Epic of the Bharata Dynasty”) one of the two Sanskrit epic poems of ancient India (the other being the Ramayana). The Mahabharata is an important source of information on the development of Hinduism between 400 BCE and 200 CE and is regarded by Hindus as both a text about dharma (Hindu moral law) and a history (itihasa, literally “that’s what happened”). Appearing in its present form about 400 CE, the Mahabharata consists of a mass of mythological and didactic material arranged around a central heroic narrative that tells of the struggle for sovereignty between two groups of cousins, the Kauravas (sons of Dhritarashtra, the descendant of Kuru) and the Pandavas (sons of Pandu). The poem is made up of almost 100,000 couplets—about seven times the length of the Iliad and the Odyssey combined—divided into 18 parvans, or sections, plus a supplement titled Harivamsha (“Genealogy of the God Hari”; i.e., of Vishnu). Although it is unlikely that any single person wrote the poem, its authorship is traditionally ascribed to the sage Vyasa, who appears in the work as the grandfather of the Kauravas and the Pandavas. The date and even the historical occurrence of the war that is the central event of the Mahabharata are much debated.

The story begins when the blindness of Dhritarashtra, the elder of two princes, causes him to be passed over in favour of his brother Pandu as king on their father’s death. A curse prevents Pandu from fathering children, however, and his wife Kunti asks the gods to father children in Pandu’s name. As a result, the god Dharma fathers Yudhishtira, the Wind fathers Bhima, Indra fathers Arjuna, and the Ashvins (twins) father Nakula and Sahadeva (also twins; born to Pandu’s second wife, Madri). The enmity and jealousy that develops between the cousins forces the Pandavas to leave the kingdom when their father dies. During their exile the five jointly marry Draupadi (who is born out of a sacrificial fire and whom Arjuna wins by shooting an arrow through a row of targets) and meet their cousin Krishna, who remains their friend and companion thereafter. Although the Pandavas return to the kingdom, they are again exiled to the forest, this time for 12 years, when Yudhishthira loses everything in a game of dice with Duryodhana, the eldest of the Kauravas.The feud culminates in a series of great battles on the field of Kurukshetra (north of Delhi, in Haryana state). All the Kauravas are annihilated, and, on the victorious side, only the five Pandava brothers and Krishna survive. Krishna dies when a hunter, who mistakes him for a deer, shoots him in his one vulnerable spot—his foot—and the five brothers, along with Draupadi and a dog who joins them (Dharma, Yudhisththira’s father, in disguise), set out for Indra’s heaven. One by one they fall on the way, and Yudhisthira alone reaches the gate of heaven. After further tests of his faithfulness and constancy, he is finally reunited with his brothers and Draupadi, as well as with his enemies, the Kauravas, to enjoy perpetual bliss.

The central plot constitutes little more than one fifth of the total work. The remainder of the poem addresses a wide range of myths and legends, including the romance of Damayanti and her husband Nala (who gambles away his kingdom just as Yudhishthira gambles away his) and the legend of Savitri, whose devotion to her dead husband persuades Yama, the god of death, to restore him to life. The poem also contains descriptions of places of pilgrimages.

Along with its basic plot and accounts of numerous myths, the Mahabharata reveals the evolution of Hinduism and its relations with other religions during its composition. The period during which the epic took shape was one of transition from Vedic sacrifice to sectarian Hinduism, as well as a time of interaction—sometimes friendly, sometimes hostile—with Buddhism and Jainism. Different sections of the poem express varying beliefs, often in creative tension. Some sections—such as the Narayaniya (a part of book 13), the Bhagavadgita (book 6), the Anugita (book 14), and the Harivamsha—are important sources of early Vaishnava theology, in which Krishna is an avatar of the god Vishnu. Above all, the Mahabharata is an exposition of dharma (codes of conduct), including the proper conduct of a king, of a warrior, of an individual living in times of calamity, and of a person seeking to attain moksha (freedom from samsara, or rebirth). The poem repeatedly demonstrates that the conflicting codes of dharma are so “subtle” that, in some situations, the hero cannot help but violate them in some respect, no matter what choice he makes.The Mahabharata story has been retold in written and oral Sanskrit and vernacular versions throughout South and Southeast Asia. Its various incidents have been portrayed in stone, notably in sculptured reliefs at Angkor Wat and Angkor Thom in Cambodia, and in Indian miniature paintings.'''

## First we will be Importing tensorflow and Tokenizer - The main reason using Tokenizer is to map each word with a number
## Example - I want apple
- I - 1, want -2 , apple - 3

In [4]:
import tensorflow as tf
from tensorflow.keras.preprocessing.text import Tokenizer

In [5]:
token = Tokenizer()
token.fit_on_texts([text])

In [8]:
# token.word_index will give us the mapped number to a word
token.word_index

{'the': 1,
 'of': 2,
 'and': 3,
 'a': 4,
 'in': 5,
 'to': 6,
 'is': 7,
 'as': 8,
 'his': 9,
 'mahabharata': 10,
 'one': 11,
 'its': 12,
 'that': 13,
 'poem': 14,
 'who': 15,
 'with': 16,
 'an': 17,
 'on': 18,
 'dharma': 19,
 'kauravas': 20,
 'their': 21,
 'for': 22,
 'pandavas': 23,
 'god': 24,
 'are': 25,
 'when': 26,
 'him': 27,
 'father': 28,
 'krishna': 29,
 'sanskrit': 30,
 'epic': 31,
 'two': 32,
 'hinduism': 33,
 'between': 34,
 'by': 35,
 'central': 36,
 'pandu': 37,
 'from': 38,
 'fathers': 39,
 'kingdom': 40,
 'during': 41,
 'five': 42,
 'draupadi': 43,
 'brothers': 44,
 'some': 45,
 'book': 46,
 'other': 47,
 'important': 48,
 '400': 49,
 'ce': 50,
 'about': 51,
 'what': 52,
 'cousins': 53,
 'sons': 54,
 'dhritarashtra': 55,
 'times': 56,
 'or': 57,
 'sections': 58,
 'vishnu': 59,
 'although': 60,
 'person': 61,
 'work': 62,
 'story': 63,
 'king': 64,
 'death': 65,
 'children': 66,
 'wife': 67,
 'pandu’s': 68,
 'arjuna': 69,
 'twins': 70,
 'also': 71,
 'born': 72,
 'dies': 7

## MAKING OUR DATA FROM INPUT TEXT
Now for each sentences we will be first split  on the basis of new line .
After that we are using texts_to_sequences method to create a list of numbers that are mapped to each number
Example - I live in Mumbai - [12,23,219,1]

finally with the help of for loop we are making our dataset which we have discussed in the readme file -
[12,23] , [12,23,219] , [12,23,219,1] for sentence 1 so that we can further split it into X and Y

In [15]:
input_sequences = []
for sentence in text.split('\n'):
  tokenized_sentence = token.texts_to_sequences([sentence])[0]

  for i in range(1, len(tokenized_sentence)):
    input_sequences.append(tokenized_sentence[:i+1])
  print(input_sequences)

[[10, 30], [10, 30, 96], [10, 30, 96, 31], [10, 30, 96, 31, 2], [10, 30, 96, 31, 2, 1], [10, 30, 96, 31, 2, 1, 97], [10, 30, 96, 31, 2, 1, 97, 98], [10, 30, 96, 31, 2, 1, 97, 98, 11], [10, 30, 96, 31, 2, 1, 97, 98, 11, 2], [10, 30, 96, 31, 2, 1, 97, 98, 11, 2, 1], [10, 30, 96, 31, 2, 1, 97, 98, 11, 2, 1, 32], [10, 30, 96, 31, 2, 1, 97, 98, 11, 2, 1, 32, 30], [10, 30, 96, 31, 2, 1, 97, 98, 11, 2, 1, 32, 30, 31], [10, 30, 96, 31, 2, 1, 97, 98, 11, 2, 1, 32, 30, 31, 99], [10, 30, 96, 31, 2, 1, 97, 98, 11, 2, 1, 32, 30, 31, 99, 2], [10, 30, 96, 31, 2, 1, 97, 98, 11, 2, 1, 32, 30, 31, 99, 2, 100], [10, 30, 96, 31, 2, 1, 97, 98, 11, 2, 1, 32, 30, 31, 99, 2, 100, 101], [10, 30, 96, 31, 2, 1, 97, 98, 11, 2, 1, 32, 30, 31, 99, 2, 100, 101, 1], [10, 30, 96, 31, 2, 1, 97, 98, 11, 2, 1, 32, 30, 31, 99, 2, 100, 101, 1, 47], [10, 30, 96, 31, 2, 1, 97, 98, 11, 2, 1, 32, 30, 31, 99, 2, 100, 101, 1, 47, 102], [10, 30, 96, 31, 2, 1, 97, 98, 11, 2, 1, 32, 30, 31, 99, 2, 100, 101, 1, 47, 102, 1], [10, 30,

## 0 PADDING TO INPUT EMBEDDING
Now we need padding as these are our embedding and all the embedding should be of same dimension. To acheive that we will be using pad_sequences function of keras which will perform 0 padding to that start (pre) of each of our list of embeddings.
To make embedding of same dim. we are finding the maximum length sentence of our text and use this as our length (dimension) of each embedding list.And this makes sense as well !

In [19]:
max_len = max([len(x) for x in input_sequences])
print(max_len)

303


In [17]:
from tensorflow.keras.preprocessing.sequence import pad_sequences

padded_input_sequences = pad_sequences(input_sequences, maxlen = max_len, padding = 'pre')

In [18]:
padded_input_sequences

array([[  0,   0,   0, ...,   0,  10,  30],
       [  0,   0,   0, ...,  10,  30,  96],
       [  0,   0,   0, ...,  30,  96,  31],
       ...,
       [  0,   0,   0, ...,   3,   5, 415],
       [  0,   0,   0, ...,   5, 415, 416],
       [  0,   0,   0, ..., 415, 416, 417]], dtype=int32)

## MAKING INPUT FEATURES AND TARGET
We will spilt our list now to X and Y, simple logic take all columns except last as X and last column as Y

In [20]:
X = padded_input_sequences[:,:-1]
Y = padded_input_sequences[:,-1]

In [21]:
X

array([[  0,   0,   0, ...,   0,   0,  10],
       [  0,   0,   0, ...,   0,  10,  30],
       [  0,   0,   0, ...,  10,  30,  96],
       ...,
       [  0,   0,   0, ..., 414,   3,   5],
       [  0,   0,   0, ...,   3,   5, 415],
       [  0,   0,   0, ...,   5, 415, 416]], dtype=int32)

In [25]:
Y

array([ 30,  96,  31,   2,   1,  97,  98,  11,   2,   1,  32,  30,  31,
        99,   2, 100, 101,   1,  47, 102,   1, 103,   1,  10,   7,  17,
        48, 104,   2, 105,  18,   1, 106,   2,  33,  34,  49, 107,   3,
       108,  50,   3,   7, 109,  35, 110,   8, 111,   4, 112,  51,  19,
       113, 114, 115,   3,   4, 116, 117, 118, 119,  52, 120, 121,   5,
        12, 122, 123,  51,  49,  50,   1,  10, 124,   2,   4, 125,   2,
       126,   3, 127, 128, 129, 130,   4,  36, 131, 132,  13, 133,   2,
         1, 134,  22, 135,  34,  32, 136,   2,  53,   1,  20,  54,   2,
        55,   1, 137,   2, 138,   3,   1,  23,  54,   2,  37,   1,  14,
         7, 139, 140,   2, 141, 142, 143, 144, 145,  56,   1, 146,   2,
         1, 147,   3,   1, 148, 149, 150, 151, 152,  57,  58, 153,   4,
       154, 155, 156, 157,   2,   1,  24, 158, 159, 160,   2,  59,  60,
       161,   7, 162,  13, 163, 164,  61, 165,   1,  14,  12, 166,   7,
       167, 168,   6,   1, 169, 170,  15, 171,   5,   1,  62,   

In [27]:
Y.shape

(813,)

In [35]:
X.shape

(813, 302)

In [29]:
len(token.word_index)

417

## ONE HOT ENCODING - TARGET VARIABLE
Now we will convert our Y into one hot encoding using to_categorical function of tensorflow. The total number of classes that are present in Y is equal to total number of words in our text. which is 417+1 as indexing starts from 1.

In [30]:
from tensorflow.keras.utils import to_categorical
y = to_categorical(Y, num_classes = 418)

In [33]:
y

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]], dtype=float32)

## DEFINING MODEL PARAMETERS

Now The main part, defining our neural network. Our NN will contain 3 layers.
1 is Embedding layer that will convert each word of our list to a vector of 100 dimension. Total words in our vocab is 418 and total number of rows are 302.
2 is LSTM layer which will output in 200 dimension
and then a dense layer which will again contain the probability of each word so the dimension should be equal to the total word in vocab and the softmax activation will give us probab for each word

While Compliling we are using categorical_crossentropy as our loss function which is used for categorical data , optimiser as adam and metrics as accuracy

In [39]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense, GRU

In [50]:
model = Sequential()
model.add(Embedding(418, 100, input_length = 302))
model.add(LSTM(200))
model.add(Dense(418, activation = 'softmax'))

In [51]:
model.compile(loss='categorical_crossentropy', optimizer = 'adam', metrics = ['accuracy'])

In [52]:
model.summary()

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_5 (Embedding)     (None, 302, 100)          41800     
                                                                 
 lstm_2 (LSTM)               (None, 200)               240800    
                                                                 
 dense_5 (Dense)             (None, 418)               84018     
                                                                 
Total params: 366618 (1.40 MB)
Trainable params: 366618 (1.40 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [53]:
model2 = Sequential()
model2.add(Embedding(418, 100, input_length = 302))
model2.add(GRU(200))
model2.add(Dense(418, activation = 'softmax'))

In [54]:
model2.compile(loss='categorical_crossentropy', optimizer = 'adam', metrics = ['accuracy'])

In [55]:
model2.summary()

Model: "sequential_6"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_6 (Embedding)     (None, 302, 100)          41800     
                                                                 
 gru_3 (GRU)                 (None, 200)               181200    
                                                                 
 dense_6 (Dense)             (None, 418)               84018     
                                                                 
Total params: 307018 (1.17 MB)
Trainable params: 307018 (1.17 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


## TRAINING
Now we are fitting our X and Y and doing forward and backward propagation 100 times so that the weights are properly optimised.

P.S -  Notice that after epoch 40 are getting same accuracy . we could use early stoppage here

In [56]:
model.fit(X,y, epochs = 100)

Epoch 1/100
26/26 [==============================] - 19s 657ms/step - loss: 5.8300 - accuracy: 0.0861
Epoch 2/100
26/26 [==============================] - 16s 620ms/step - loss: 5.4430 - accuracy: 0.0923
Epoch 3/100
26/26 [==============================] - 16s 619ms/step - loss: 5.3570 - accuracy: 0.0923
Epoch 4/100
26/26 [==============================] - 17s 659ms/step - loss: 5.3237 - accuracy: 0.0923
Epoch 5/100
26/26 [==============================] - 16s 620ms/step - loss: 5.2872 - accuracy: 0.0923
Epoch 6/100
26/26 [==============================] - 16s 619ms/step - loss: 5.2298 - accuracy: 0.0984
Epoch 7/100
26/26 [==============================] - 17s 656ms/step - loss: 5.1442 - accuracy: 0.1058
Epoch 8/100
26/26 [==============================] - 16s 617ms/step - loss: 5.0252 - accuracy: 0.1328
Epoch 9/100
26/26 [==============================] - 16s 618ms/step - loss: 4.8423 - accuracy: 0.1611
Epoch 10/100
26/26 [==============================] - 17s 660ms/step - loss: 4.610

## PREDICTION
Finally we are testing our model by predicting next word 10 times .
Ofcourse first we need to map our text to a number and then 0 pad it and then pass it to the model

In [62]:
import numpy as np

pred_text = 'Hindu'

for i in range(10):
  token_pred_text = token.texts_to_sequences([pred_text])[0]
  padded_token_pred_text = pad_sequences([token_pred_text], maxlen = 302, padding = 'pre')
  pos = np.argmax(model.predict(padded_token_pred_text))

  for word,index in token.word_index.items():
    if index == pos:
      pred_text= pred_text+ ' '+ word
      print(pred_text)

1/1 [==============================] - 0s 129ms/step
Hindu “great
1/1 [==============================] - 0s 82ms/step
Hindu “great plot
1/1 [==============================] - 0s 74ms/step
Hindu “great plot constitutes
1/1 [==============================] - 0s 79ms/step
Hindu “great plot constitutes little
1/1 [==============================] - 0s 79ms/step
Hindu “great plot constitutes little more
1/1 [==============================] - 0s 78ms/step
Hindu “great plot constitutes little more than
1/1 [==============================] - 0s 79ms/step
Hindu “great plot constitutes little more than one
1/1 [==============================] - 0s 81ms/step
Hindu “great plot constitutes little more than one fifth
1/1 [==============================] - 0s 75ms/step
Hindu “great plot constitutes little more than one fifth of
1/1 [==============================] - 0s 77ms/step
Hindu “great plot constitutes little more than one fifth of the
